### PR TITLE
When `spin watch` terminates spin up, allow it to delete its temp dir

### LINKED ISSUE #1698: What is the recommended solution to clear spin-related cache files in the /tmp directory?
Loving spin otherwise btw!

I've been using spin locally via `spin watch` during development of an app using 3 components (the spin-fileserver and 2 custom ones) on a Debian Linux system for weeks. Noticed today my `/tmp` partition was nearly out of space. Seeing thousands of directories called `.tmpRANDOM` (where `RANDOM` is a random 6-digit alphanumeric string) with `spin.lock` (so I know they're related to spin) in `/tmp` totaling to hundreds of GB!

Not observing any subcommand in the `spin` CLI addressing cleaning up cache, etc...

Presuming I can kill `spin watch` and any other instances of spin, and delete all of these via `fd '^spin.lock$' /tmp |sed 's/\/spin\.lock$//' |xargs rm -rf`?

Is there any other solution you would recommend?

Are there any methods to prevent this behavior... such as... spin automatically deleting cache files in /tmp when it exits?

Thanks!

---

Spin version (`spin --version`): `spin 1.4.1 (e0bd911 2023-07-12)`

Installed plugin versions (`spin plugins list --installed`)

```
cloud 0.1.2 [installed]
js2wasm 0.5.1 [installed]
py2wasm 0.3.0 [installed]
```

### PR DESCRIPTION
**Pull request by itowlson**
Fixes #1698.

A previous crack at this in #1711 ran aground on sychronising the `spin build` and `spin up` processes when `spin up` was terminating asynchronously.  I couldn't get the watch state to remain in sync with what `watchexec` thought it was doing, resulting in either missed file changes or missed restarts.

This is, I'm afraid, a more dramatic rework, and therefore carries a higher risk - the more so because, unlike the original implementation, it does not abstract notifications or defunctionalise process control, and so it way less amenable to unit testing.  I'm not sure how to address that - I'd welcome ideas.  On the plus side, the individual components at work here are rather more simple-minded, so _hopefully_ less in need of unit testing once the inevitable initial bugs have been shaken out.

The change in strategy is twofold:

1. Instead of a single `watchexec` watching all interesting files, and patching up "what command is watchexec running" depending on what files have changed, this now uses two separate watchexecs, one watching artifact files, one watching build files.
2. These watchexecs do not manage processes themselves. Instead, they send notifications to two processor tasks, the _buildifier_ and _uppificator_.  (Okay, yes.  I should probably change the names.)  These spawn and restart processes according to the notifications they receive, and coordinate so that changes triggered by a build do not start `spin up` until the build is complete.

This means that the processor tasks are much less dependent on state than the previous implementation - they are pretty much simple loops going round and round waiting for the change notifications they're interested in and starting or restarting a child process.  (The only piece of state that needs to be passed between notifications is "don't clear the screen because build just did".)

This has so far been tested only on a small Rust application.  Next step is to try it on something like the docs site, where the build process makes much more far-reaching changes to the artifacts.  It should also adopt some of the "stupid operating system does what" patches from the previous implementation!

In the meantime I'm putting this out for folks to kick the tyres and yell at the design/implementation.  @calebschoepp needless to say, I would particularly value your experience and insight on the practicalities as well as the design!

### CODE REVIEW

**Comment by itowlson** [comment]
Well there's a bug where the docs site makes it go into an infinite loop so we're off to a flying start.

**Comment by calebschoepp** [comment]
@itowlson I'll take a look at this when I get a chance, but I won't have time today. Thanks for diving so deep on `spin watch`.

File: src/commands/watch/uppificator.rs
Comment (calebschoepp) [inline src/commands/watch/uppificator.rs:90]: Now that we're manually doing the process management and killing what benefit are we getting from `watchexec`? Did Notify suite our needs? If we could it'd be nice to get rid of a dependency.

**Review by calebschoepp — CHANGES_REQUESTED** [review]
Great work @itowlson! I think we're headed in the right direction

File: Cargo.toml
Comment (itowlson) [reply Cargo.toml:21]: I believe so.  We call `clearscreen::clear()` to clear the screen on a change.

File: Cargo.toml
Comment (itowlson) [reply Cargo.toml:37]: Nope - thanks for the catch!

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:128]: Since it's not a regression, I'll defer it to a followup PR.  That minimises the complexity of this one and keeps it as a pure bug fix of the existing functionality, so that people can assess the impact of the re-read feature separately.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:139]: I can't promise zero but I'm confident saying that its overhead is insignificant compared to all the other clanking, wheezing, hissing, and venting of steam going on in this process!  There is, I think, some small cost setting up the sleep and coming out of it, but unless you're looping a sleep with a duration in the millisecond range it's ignorable.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:188]: The existing `spin watch` code has it at debug level I believe.  I'm happy to leave it there - the reason I changed it was that some folks felt it should be more visible - let me know if you feel strongly it should be put back to debug.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:197]: Hmm, it means `JoinHandle` but `_handle` feels too ambiguous, and `_join_handle` feels quite long.  Changed it to `join_handle` since it's only used on two lines and may the `cargo fmt` gods have mercy on my soul.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:203]: Yes the existing ignore stuff is definitely something I skipped while trying to get this working, and wanted to understand from you.  My impression was that the existing implementation has only one ignore path, `*.swp`, and with the dual-watchexec thing it would be applied to both filters, correct?  I'm not sure what this pattern is though, so if you have time to walk me through it, that would be super useful - thanks!

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:161]: When any or the tasks (up manager, build manager, artifact watch, build source watch) exit, the the `futures::future::select_all` fires.  At that point `Watch::run()` returns and all remaining tasks go down with the process.  The usual exit route is user terminates with Ctrl+C, that fires a stop message to the Uppificator, which stops `spin up` and breaks out of its loop, ending its task and thus causing `select_all` to return and the process to exit.  It sounds like I need to be more explicit about this in the documentation!

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:176]: Hmm, most of this stuff uses the Watch command object.  We could capture the relevant bits of configuration and pass them explicitly, but I'm not sure of the value of it.  Definitely something to _arf arf_ watch though... I agree this is a pretty dense file...

Comment (itowlson) [reply src/commands/watch.rs:176]: In terms of system notify limits you mean?  That's a reasonable concern for sure, and one I probably wouldn't pick up in simple test cases.  I did try it on the docs site but that's still medium-scale rather than enormous.  The default notify limit is, I think, 8192.  What do you reckon?

File: src/commands/watch/buildifier.rs
Comment (itowlson) [reply src/commands/watch/buildifier.rs:10]: We use the `watch` channel because the receivers don't actually care which file has changed, and they don't need to respond to every change: they care only only that _something_ in their area of interest has changed since the last time they checked.  "Has something changed" is the exact use case for the `watch` channel, which does not transmit a sequence of values but holds just one mutable value; its `receive` returns when the value is different from the last `receive`.  This saves us from having to receive our way through a queue of multiple changes.

Using a `Uuid` means we don't need to worry about concocting a unique "what changed" message.  For example, if we sent the filename, then we couldn't use `watch` channels because if `fie.txt` changed (causing a reload) and then `fie.txt` changed again, that would _not_ be a change to the value in the channel.  `Uuid` guarantees that a new value will be generated for every debounced change set.

File: src/commands/watch/buildifier.rs
Comment (itowlson) [reply src/commands/watch/buildifier.rs:28]: Good point.  Although I'm slightly wary of cancelling builds after some bad experiences locking myself out of a Rust build because of interrupting it as a vital point.  But my wariness is not a good reason for running a lengthy build to completion just to run it all over again.  Time for some more select-related goodness I guess!

File: src/commands/watch/uppificator.rs
Comment (itowlson) [reply src/commands/watch/uppificator.rs:20]: Uh... I don't know... I tend to do it when I need it.  Though there's no harm in doing it pre-emptively in case someone else - some kind, generous, forgiving Canadian for example - later needs to debug one's awful code and would prefer not to have to whack the derive on themself.

File: src/commands/watch/uppificator.rs
Comment (itowlson) [reply src/commands/watch/uppificator.rs:63]: Correct.

File: src/commands/watch/uppificator.rs
Comment (itowlson) [reply src/commands/watch/uppificator.rs:81]: It could, but I left it running so people could continue testing while waiting for the build to complete.  The idea was to minimise downtime, but maybe it's confusing.

File: src/commands/watch/uppificator.rs
Comment (itowlson) [reply src/commands/watch/uppificator.rs:90]: `watchexec` layers filtering on top of `notify` and I couldn't see a way to use filterers with raw notify.  (My attempt to use raw notify is why that leftover dependency.)  If we can do that then I'm totally up for it.

File: src/commands/watch/buildifier.rs
Comment (itowlson) [reply src/commands/watch/buildifier.rs:28]: Hmm, so I tried simulating a Rust program taking a long time to build by having `build.rs` that just slept for 5 seconds.  (I know.  I had to _simulate_ long Rust builds.  You _failed_ me, Rust.)  And cancelling the build just... didn't work.  Neither kill or interrupt stopped it.

I'm hoping this was a weird artifact of using the `build.rs` kludge, because I _know_ I've cancelled Rust builds before.  Or maybe `spin build` was getting killed but not propagating the cancellation to `cargo build`?

Comment (itowlson) [reply src/commands/watch/buildifier.rs:28]: I think `spin build` might not be propagating the cancellation.  It looks like there's a shell between `spin build` and the actual build process which won't help.

(This is the processes running after I made a change, saved it, made another change, and saved that - notice only one `spin build` but two `cargo build`.  There was also a `cargo check` but I believe that was rust-analyzer catching up with my edits.)

```
ivan@hecate:~$ ps -A -f | egrep "(spin)|(cargo)"
ivan      9724 26324  0 16:32 pts/17   00:00:00 spin watch
ivan     10021  9724  0 16:33 pts/17   00:00:00 /home/ivan/github/spin/target/debug/spin up -f spin.toml
ivan     10034 10021  5 16:33 pts/17   00:00:04 /home/ivan/github/spin/target/debug/spin trigger http
ivan     10543 26323  0 16:34 pts/17   00:00:00 sh -c cargo build --target wasm32-wasi --release
ivan     10544 10543  0 16:34 pts/17   00:00:00 /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/cargo build --target wasm32-wasi --release
ivan     10606  9724  0 16:34 pts/17   00:00:00 /home/ivan/github/spin/target/debug/spin build -f spin.toml
ivan     10616 10606  0 16:34 pts/17   00:00:00 sh -c cargo build --target wasm32-wasi --release
ivan     10617 10616  2 16:34 pts/17   00:00:00 /home/ivan/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/cargo build --target wasm32-wasi --release
```

Comment (itowlson) [reply src/commands/watch/buildifier.rs:28]: The existing `spin watch` handles this correctly so I guess `watchexec` is killing things more effectively than I am!  I'll have a look and see if I can see what it does.

Comment (itowlson) [reply src/commands/watch/buildifier.rs:28]: Looks like some magic deep in the bowels of `command_group` which is hopefully reusable.

Comment (itowlson) [reply src/commands/watch/buildifier.rs:29]: Okay I believe this is now working

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs]: Sounds good to me

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:202]: This is a strong opinion loosely held situation.

I'm sure that _my opinion_™ is that it should be at the `debug` level. I'm of the opinion that it's really noisy and you'd only really want this information when you're debugging the plugin as a developer of Spin.

That being said, my strong opinion is loosely held so I'll leave it up to you to make a judgement call of what you think is right. We can always change it later anyways 👍

File: src/commands/watch.rs
Comment (calebschoepp) [inline src/commands/watch.rs:338]: Mega nit: I know we're in the land of made up terms here but this is how I would spell it.
```suggestion
fn stringize_path(path: &Path) -> anyhow::Result<String> {
```

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:139]: > clanking, wheezing, hissing, and venting of steam
Your technical wizardry has sufficiently calmed my anxieties 🤣

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:217]: You are correct that the original implementation only had the one ignore path. I originally planned to not support ignore paths but ran into issues when I was testing with Vim where the swap files it would create would cause a deluge  of restarts. My very hack solution was to introduce this very naive ignore list to the filter with only the single path for now. It's hack because we don't expose this functionality to the users at all and because the pattern doesn't actually cover all the possible swap files Vim could make (see [here](https://unix.stackexchange.com/questions/326707/vim-what-are-all-the-possible-swapfile-extensions)).

To maintain backwards compatibility I think we should support ignoring any files matching `*.swp`. Going forward I think it would be good to expose this through the `spin.toml` configuration so users can get more granular about what they want to ignore. Perhaps through gitignore syntax like `!*.swp`.

As to whether this should be applied to both filters — that is a good question. My gut reaction is that yes it should be applied to both. For example if you were editing a static asset with Vim it would create a swap file that would match the static assets. Imagining the future where we support user provided ignore paths it is less clear to me if they should be added to both filters or not. That's maybe an issue we can address at that point.

If you need more explanation ping me and we can chat about it.

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:164]: Yep, documenting that would be great!

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:190]: Good point on it being reliant on the Watch command object. That's a great reason to keep it all together for now.

Comment (calebschoepp) [reply src/commands/watch.rs:190]: Honestly I meant in terms of other system resources like CPU, memory which is probably far from any real concern. Good point on the notify limits. I'd say we're probably handling that when it actually becomes a problem.

File: src/commands/watch/buildifier.rs
Comment (calebschoepp) [reply src/commands/watch/buildifier.rs:11]: Thanks for the education 😄

File: src/commands/watch/uppificator.rs
Comment (calebschoepp) [reply src/commands/watch/uppificator.rs:91]: Let's just punt on it then. It seems like it is still providing us enough value.

File: src/commands/watch/uppificator.rs
Comment (calebschoepp) [reply src/commands/watch/uppificator.rs:82]: I didn't think about the downtime reduction. I like that. Let's keep it like this for now.

**Review by calebschoepp — COMMENTED** [review]
I responded to your comments and took another pass over your changes. I think we're getting close. The only stuff that I really feel is still blocking is the ignore list work.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:202]: I'll change it back to `debug`.  Can't please everybody and addressing the "make it more visible" request is out of scope for this PR.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs:338]: oops, thanks

**Comment by itowlson** [comment]
@calebschoepp I think this addresses everything.  I have left the commit history intact for now so you can see changes since first review if that's useful; if approved, I'll squash before merging (well if I remember anyway).

**Review by calebschoepp — APPROVED** [review]
Great work @itowlson , LGTM! Ship it 🚢

**Comment by itowlson** [comment]
@calebschoepp I implemented your fix for the Mac source notification issue.  It Works On My Machine(s) TM, but if you have time to test it on yours, that would greatly increase my confidence.  Thanks!

**Comment by calebschoepp** [comment]
> @calebschoepp I implemented your fix for the Mac source notification issue. It Works On My Machine(s) TM, but if you have time to test it on yours, that would greatly increase my confidence. Thanks!

I built latest on this branch and we no longer get the infinite looping behaviour. I'm seeing one infinite loop bug persist though. I sent you a video of it in Slack.

**Pull request merged**
